core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -310,15 +305,33 @@ public void registerFunctionScoreParser(ScoreFunctionParser<? extends ScoreFunct
      *        registered. So it is the name that the query should use as its {@link NamedWriteable#getWriteableName()}.
      */
     public <QB extends QueryBuilder<QB>> void registerQuery(Writeable.Reader<QB> reader, QueryParser<QB> parser, String... names) {
```

Comment:
I agree with you 100% we have to be more careful with query ids and serialization problems. With this change we hide even more that the first name is used for serialization, maybe we can already split the two (parser names and writeable id) more clearly instead, I would go that path.